future-apps/agent/contracts/SignatureValidator.sol

```diff
@@ -0,0 +1,137 @@
+pragma solidity 0.4.24;
+
+// Inspired in https://github.com/horizon-games/multi-token-standard/blob/master/contracts/utils/SignatureValidator.sol
+// This should probably be moved into aOS: https://github.com/aragon/aragonOS/pull/442
+
+import "./standards/ERC1271.sol";
+
+
+library SignatureValidator {
+    enum SignatureMode {
+        Invalid, // 0x00
+        EIP712,  // 0x01
+        EthSign, // 0x02
+        ERC1271, // 0x03
+        NMode    // 0x04, to check if mode is specified, leave at the end
+    }
+
+    // bytes4(keccak256("isValidSignature(bytes,bytes)")
+    bytes4 constant public ERC1271_RETURN_VALID_SIGNATURE = 0x20c13b0b;
+    uint256 internal constant ERC1271_ISVALIDSIG_MAX_GAS = 250000;
+
+    /// @dev Validates that a hash was signed by a specified signer.
+    /// @param hash Hash which was signed.
+    /// @param signer Address of the signer.
+    /// @param signature ECDSA signature along with the mode (0 = Invalid, 1 = EIP712, 2 = EthSign, 3 = ERC1271) {mode}{r}{s}{v}.
+    /// @return Returns whether signature is from a specified user.
+    function isValidSignature(bytes32 hash, address signer, bytes signature) internal view returns (bool) {
+        if (signature.length == 0) {
+            return false;
+        }
+
+        uint8 modeByte = uint8(signature[0]);
+        if (modeByte >= uint8(SignatureMode.NMode)) {
+            return false;
+        }
+        SignatureMode mode = SignatureMode(modeByte);
+
+        if (mode == SignatureMode.EIP712) {
+            return ecVerify(hash, signer, signature);
+        } else if (mode == SignatureMode.EthSign) {
+            return ecVerify(
+                keccak256(abi.encodePacked("\x19Ethereum Signed Message:\n32", hash)),
+                signer,
+                signature
+            );
+        } else if (mode == SignatureMode.ERC1271) {
+            // Pop the mode byte before sending it down the validation chain
```

Comment:
`if (signature.length <= 1) return false`

Reply:
Not needed, see https://github.com/aragon/aragon-apps/pull/580#discussion_r257318004